File: extension/vscode/src/code-api.ts
import * as vscode from 'vscode';

export type BreakpointEditResult = { newCode: string; breakpointLine: number | null };
export type CodeEditResult = { newCode: string };

async function postCodeEdit<T>(
  serverUrl: string,
  endpoint: string,
  body: Record<string, unknown>,
  logger: vscode.OutputChannel,
): Promise<T | null> {
  if (!serverUrl) {
    return null;
  }
  try {
    const response = await fetch(`${serverUrl}/api/code/${endpoint}`, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify(body),
    });
    if (!response.ok) {
      logger.appendLine(`[code-api] ${endpoint} returned HTTP ${response.status}`);
      return null;
    }
    return (await response.json()) as T;
  } catch (err) {
    logger.appendLine(`[code-api] ${endpoint} failed: ${err}`);
    return null;
  }
}

export function addBreakpoint(serverUrl: string, code: string, referenceRow: number, logger: vscode.OutputChannel) {
  return postCodeEdit<BreakpointEditResult>(serverUrl, 'add-breakpoint', { code, referenceRow }, logger);
}

export function removeBreakpoint(serverUrl: string, code: string, line: number, logger: vscode.OutputChannel) {
  return postCodeEdit<BreakpointEditResult>(serverUrl, 'remove-breakpoint', { code, line }, logger);
}

export function toggleBreakpoint(serverUrl: string, code: string, cursorRow: number, logger: vscode.OutputChannel) {
  return postCodeEdit<BreakpointEditResult>(serverUrl, 'toggle-breakpoint', { code, cursorRow }, logger);
}

export function clearBreakpoints(serverUrl: string, code: string, logger: vscode.OutputChannel) {
  return postCodeEdit<CodeEditResult>(serverUrl, 'clear-breakpoints', { code }, logger);
}

export function insertPoint(
  serverUrl: string, code: string, sourceLine: number, point: [number, number], logger: vscode.OutputChannel,
) {
  return postCodeEdit<CodeEditResult>(serverUrl, 'insert-point', { code, sourceLine, point }, logger);
}

export function removePoint(
  serverUrl: string, code: string, sourceLine: number, point: [number, number], logger: vscode.OutputChannel,
) {
  return postCodeEdit<CodeEditResult>(serverUrl, 'remove-point', { code, sourceLine, point }, logger);
}

export function addPick(serverUrl: string, code: string, sourceLine: number, logger: vscode.OutputChannel) {
  return postCodeEdit<CodeEditResult>(serverUrl, 'add-pick', { code, sourceLine }, logger);
}

export function setPickPoints(
  serverUrl: string, code: string, sourceLine: number, points: [number, number][], logger: vscode.OutputChannel,
) {
  return postCodeEdit<CodeEditResult>(serverUrl, 'set-pick-points', { code, sourceLine, points }, logger);
}

/**
 * Replace the entire contents of `doc` with `newCode` in a single workspace
 * edit. Returns true on success.
 */
export async function replaceDocument(doc: vscode.TextDocument, newCode: string): Promise<boolean> {
  if (doc.getText() === newCode) {
    return true;
  }
  const fullRange = new vscode.Range(
    new vscode.Position(0, 0),
    doc.lineAt(doc.lineCount - 1).range.end,
  );
  const edit = new vscode.WorkspaceEdit();
  edit.replace(doc.uri, fullRange, newCode);
  return vscode.workspace.applyEdit(edit);
}

File: extension/neovim/lua/fluidcad/code_api.lua
local M = {}

local function get_url()
  local ok, bridge = pcall(require, 'fluidcad.bridge')
  if not ok then
    return nil
  end
  return bridge.get_url()
end

--- POST a JSON body to /api/code/<endpoint> and return the decoded response,
--- or nil on any failure (server unreachable, non-2xx, decode error).
local function post(endpoint, body)
  local url = get_url()
  if not url or url == '' then
    return nil
  end
  local encoded = vim.fn.json_encode(body)
  local result = vim.fn.system(
    { 'curl', '-s', '-f', '-X', 'POST',
      '-H', 'Content-Type: application/json',
      '--data-binary', '@-',
      url .. '/api/code/' .. endpoint },
    encoded
  )
  if vim.v.shell_error ~= 0 then
    return nil
  end
  local ok, decoded = pcall(vim.fn.json_decode, result)
  if not ok or type(decoded) ~= 'table' then
    return nil
  end
  return decoded
end

function M.add_breakpoint(code, reference_row)
  return post('add-breakpoint', { code = code, referenceRow = reference_row })
end

function M.remove_breakpoint(code, line)
  return post('remove-breakpoint', { code = code, line = line })
end

function M.toggle_breakpoint(code, cursor_row)
  return post('toggle-breakpoint', { code = code, cursorRow = cursor_row })
end

function M.clear_breakpoints(code)
  return post('clear-breakpoints', { code = code })
end

function M.insert_point(code, source_line, point)
  return post('insert-point', { code = code, sourceLine = source_line, point = point })
end

function M.remove_point(code, source_line, point)
  return post('remove-point', { code = code, sourceLine = source_line, point = point })
end

function M.add_pick(code, source_line)
  return post('add-pick', { code = code, sourceLine = source_line })
end

function M.set_pick_points(code, source_line, points)
  return post('set-pick-points', { code = code, sourceLine = source_line, points = points })
end

--- Replace the entire contents of `bufnr` with `new_code`. Returns true on
--- success.
function M.replace_buffer(bufnr, new_code)
  if not vim.api.nvim_buf_is_valid(bufnr) then
    return false
  end
  local lines = vim.split(new_code, '\n', { plain = true })
  vim.api.nvim_buf_set_lines(bufnr, 0, -1, false, lines)
  return true
end

return M
